arch/xtensa/core/xtensa_asm2_util.S

```diff
@@ -160,6 +160,19 @@ _restore_context:
 	FPU_REG_RESTORE
 #endif
 
+#if (XCHAL_CP_MASK & (1 << XCHAL_CP_ID_AUDIOENGINELX)) && \
+	defined(CONFIG_CPU_HAS_DSP) && defined(CONFIG_DSP_SHARING)
+
+	/* Put aligned address of ae_regs into A2 */
+
+	addi a2, a1, ___xtensa_irq_bsa_t_ae_regs_OFFSET
+	movi a3, ~(XCHAL_CP1_SA_ALIGN - 1)
+	addi a2, a2, (XCHAL_CP1_SA_ALIGN - 1)
+	and a2, a2, a3
+
+	xchal_cp1_load a2 a0 a3 a3 a3    /* Only A0 and A2 are used by macro */
```

Comment:
As mentioned, really really not a fan of seeing HAL code here in the core entry code, especially since we all agree we're going to throw out this technique in the future anyway and move to a lazy-evaluated version based on the CP enable trap.

If we really have to merge this version first, can we move it to a separate .S file where we can isolate the HAL pollution and call it as a subroutine (or just a CALL0 I guess, since there's no stack needed and the registers get clobbered anyway)?

Reply:
Isolated.